openedx/core/djangoapps/credit/signals.py

```diff
@@ -70,3 +99,264 @@ def listen_for_grade_calculation(sender, username, grade_summary, course_key, de
+    """ Create verification user partition for given block.
+
+    Group and UserPartition id's will be int.
+
+    Args:`
+        course_key (CourseKey): Identifier for the course
+        user_partition (UserPartition): UserPartition object
+        block (XBlock): XBlock mixin
+
+    Returns:
+        UserPartition object.
+
+    """
+    course = modulestore().get_course(course_key, depth=0)
+
+    # make int id for user partition scheme `VerificationPartitionScheme` which
+    # is unique in user partitions of the provided course
+    group_configuration_id = generate_int_id(MINIMUM_GROUP_ID, MYSQL_MAX_INT, get_course_partitions_used_ids(course))
+    group_configuration_name = u"Verification Checkpoint for {checkpoint_name}".format(
+        checkpoint_name=block.display_name
+    )
+    group_configuration_description = group_configuration_name
+    # Since Group requires id to be an int so map verification partition group names to int
+    # 0: VerificationPartitionScheme.NON_VERIFIED
+    # 1: VerificationPartitionScheme.VERIFIED_ALLOW
+    # 2: VerificationPartitionScheme.VERIFIED_DENY
+    group_configuration_groups = [
+        Group(VerificationPartitionScheme.NON_VERIFIED, 'Not enrolled in a verified track'),
+        Group(VerificationPartitionScheme.VERIFIED_ALLOW, 'Enrolled in a verified track and has access'),
+        Group(VerificationPartitionScheme.VERIFIED_DENY, 'Enrolled in a verified track and does not have access'),
+    ]
+    group_configuration_parameters = {'location': unicode(block.location)}
+
+    group_configuration = UserPartition(
+        id=group_configuration_id,
+        name=group_configuration_name,
+        description=group_configuration_description,
+        groups=group_configuration_groups,
+        scheme=user_partition,
+        parameters=group_configuration_parameters,
+    )
+    return group_configuration
+
+
+def _get_credit_xblocks_for_access_control(course_key):  # pylint: disable=invalid-name
+    """ Retrieve all credit requirements XBlocks in the course for categories
+     in list 'GATED_CREDIT_XBLOCK_CATEGORIES'.
+
+    Args:
+        course_key (CourseKey): Identifier for the course
+
+    Returns:
+        List of XBlocks that are published and haven't been deleted.
+
+    """
+    credit_blocks = []
+    for category in GATED_CREDIT_XBLOCK_CATEGORIES:
+        xblocks = get_course_xblocks(course_key, category)
+        credit_blocks.extend(xblocks)
+
+    return credit_blocks
+
+
+def _update_published_block(block):
+    """ Update the provided XBlock in modulestore for published branch only.
+
+    Use the xblock's last editor as modifier.
+
+    Args:
+        block (XBlock): XBlock mixin
+
+    Returns:
+        List of XBlocks that are published and haven't been deleted.
+
+    """
+    # Update the published branch only
+    with modulestore().branch_setting(ModuleStoreEnum.Branch.published_only):
```

Comment:
@doctoryes @jimabramson what I am thinking is that:
Lets assume that user made some changes for a block which now have two versions `v1` (published) and `v2` (draft) which have user's changes.
1. get the draft version of block `v2` and save it in some temporary placeholder, lets say `temp_draft`
2. made changes e.g., update `group_access` field of published version `v1` and make a new version `v3` on draft branch with `update_item`
3. publish the new version `v3`. now publish branch and draft branch will both point to new version `v3`
4. again call `update_item` with `temp_draft`, which will create new item `v4` (with saved user changes).
Now we have saved the `group_access` field changes on published branch `v3` and we also saved user changes on draft branch `v4`.
Does this make sense?
Also I will be away for this week. Please tag @awais786 and @aamir-khan If you have any questions.
FYI @wedaly @jimabramson @awais786 @aamir-khan @tasawernawaz @cahrens @andy-armstrong